tests/ha_test.go

```diff
@@ -0,0 +1,147 @@
+	}
+	suite.operator = utils.SparkOperatorInstallation{
+		Params: suite.leaderElectionParams,
+	}
+
+	if err := suite.operator.InstallSparkOperator(); err != nil {
+		suite.FailNow(err.Error())
+	}
+
+}
+
+func (suite *HighAvailabilityTestSuite) TearDownSuite() {
+	suite.operator.CleanUp()
+}
+
+func (suite *HighAvailabilityTestSuite) TestParameters() {
+	operator := suite.operator
+	params := suite.leaderElectionParams
+	args, err := utils.Kubectl("get", "deployment", operator.InstanceName, "-n", operator.Namespace,
+		"-o=jsonpath={.spec.template.spec.containers[0].args}")
+	if err != nil {
+		suite.FailNow(err.Error())
+	}
+	availableReplicas, _ := utils.Kubectl("get", "deployment", operator.InstanceName, "-n", operator.Namespace,
+		"-o=jsonpath={.status.availableReplicas}")
+
+	suite.Equal(params["replicas"], availableReplicas)
+	suite.Contains(args, fmt.Sprint("-leader-election=", params["enableLeaderElection"]))
+	suite.Contains(args, fmt.Sprint("-leader-election-lock-name=", params["leaderElectionLockName"]))
+	suite.Contains(args, fmt.Sprint("-leader-election-lock-namespace=", operator.Namespace))
+	suite.Contains(args, fmt.Sprint("-leader-election-lease-duration=", params["leaderElectionLeaseDuration"]))
+	suite.Contains(args, fmt.Sprint("-leader-election-renew-deadline=", params["leaderElectionRenewDeadline"]))
+	suite.Contains(args, fmt.Sprint("-leader-election-retry-period=", params["leaderElectionRetryPeriod"]))
+}
+
+func (suite *HighAvailabilityTestSuite) TestLeaderElection() {
+	leaderElectionRecord, err := getLeaderElectionRecord(suite.operator)
+	if suite.NoError(err) {
+		suite.NotEmpty(leaderElectionRecord.HolderIdentity)
+	}
+}
+
+func (suite *HighAvailabilityTestSuite) TestFailover() {
+	operator := suite.operator
+	leaderElectionRecord, err := getLeaderElectionRecord(operator)
+	if err != nil {
+		suite.FailNow(err.Error())
+	}
+	fmt.Println("Current leader: ", leaderElectionRecord.HolderIdentity)
+	// deploy workload
+	jobName := "mock-task-runner"
+	mockTaskRunner := utils.SparkJob{
+		Name:     jobName,
+		Template: "spark-mock-task-runner-job.yaml",
+		Params: map[string]interface{}{
+			"args": []string{"1", "30"},
+		},
+	}
+	if err := operator.SubmitJob(&mockTaskRunner); err != nil {
+		suite.FailNow(err.Error())
+	}
+	// wait until the application is running
+	if err := operator.WaitForJobState(mockTaskRunner, v1beta2.RunningState); err != nil {
+		suite.FailNow(err.Error())
+	}
+
+	log.Infof("deleting current leader pod \"%s\"", leaderElectionRecord.HolderIdentity)
+	err = utils.DeleteResource(operator.Namespace, "pod", leaderElectionRecord.HolderIdentity)
+
+	// check re-election
+	if err := utils.RetryWithTimeout(electionRecordRetryTimeout, electionRecordRetryInterval, func() error {
+		if newLeaderElectionRecord, err := getLeaderElectionRecord(operator); err != nil {
+			return err
+		} else if newLeaderElectionRecord.HolderIdentity == leaderElectionRecord.HolderIdentity {
```

Comment:
is there a possibility that the last leader can be elected again? this can expose the test to timing issues and make it flaky. Does it make sense to assign lower values to election-related time settings in parameters?

Reply:
I would say no because we delete the leader pod and the new pod is created instead by `ReplicaSet`. This condition is required to check that the leader election record is updated with the newly elected leader and force the method to retry if `HolderIdentity`  isn't changed.